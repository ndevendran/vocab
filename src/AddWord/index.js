import React, {Component} from 'react';
import {InputGroup, Button, FormControl} from 'react-bootstrap';

export default class AddWord extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      def: ''
    }

    this.nameOnChange = this.nameOnChange.bind(this);
    this.defOnChange = this.defOnChange.bind(this);
    this.addWord = this.addWord.bind(this);
  }

  nameOnChange(event) {
    this.setState({ name: event.target.value });
  }

  defOnChange(event) {
    this.setState({ def: event.target.value });
  }

  addWord() {
    const name = this.state.name;
    const def = this.state.def;
    this.props.onClick(name, def);
    this.setState({
      name: '',
      def: ''
    });
  }

  render () {
    return (
      <div className="addWord">
        {this.props.hideStep
          ? <div></div>
          :
          <div>
          <InputGroup className="mb-3 wordInput">
            <InputGroup.Prepend>
              <InputGroup.Text>Name</InputGroup.Text>
            </InputGroup.Prepend>
            <FormControl
              onChange={this.nameOnChange}
              value={this.state.name}
            />
          </InputGroup>
            <br />
            <InputGroup className="definitionInput">
              <InputGroup.Prepend>
                <InputGroup.Text>Definition</InputGroup.Text>
              </InputGroup.Prepend>
              <FormControl as="textarea"
                onChange={this.defOnChange}
                value={this.state.def}
              />
            </InputGroup>
            <br />
            <Button variant="primary" onClick={this.addWord}>Add Word</Button>
          </div>
        }
      </div>
    );
  }
}
