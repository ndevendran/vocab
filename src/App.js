import React, {Component} from 'react';
import logo from './logo.svg';
import './App.css';

const list = [
  { name: "malevolent",
    definition: "having or showing a wish to do evil to others"},
  { name: "pejorative",
    definition: "expressing contempt or disapproval"}
];

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      words: list,
      index: 0,
      hideMeaning: true,
      addWordName: '',
      addWordDef: '',
    }

    this.getNextWord = this.getNextWord.bind(this);
    this.getPreviousWord = this.getPreviousWord.bind(this);
    this.showDefinition = this.showDefinition.bind(this);
    this.nameOnChange = this.nameOnChange.bind(this);
    this.defOnChange = this.defOnChange.bind(this);
    this.addWord = this.addWord.bind(this);
  }

  getNextWord() {
    const currentIndex = this.state.index;
    const nextIndex = currentIndex + 1;
    const words = this.state.words;
    const len = words.length;

    if(nextIndex >= len) {
      this.setState({index: 0, hideMeaning: true});
    } else {
      this.setState({index: nextIndex, hideMeaning: true});
    }
  }

  getPreviousWord() {
    const previousIndex = this.state.index-1;
    const words = this.state.words;
    const len = words.length;
    if(previousIndex <= 0) {
      this.setState({ index: 0, hideMeaning: true});
    } else {
      this.setState({index: previousIndex, hideMeaning: true});
    }
  }

  showDefinition() {
    const hideMeaning = !this.state.hideMeaning;
    this.setState({hideMeaning: hideMeaning});
  }

  nameOnChange(event) {
    this.setState({ addWordName: event.target.value });
  }

  defOnChange(event) {
    this.setState({ addWordDef: event.target.value });
  }

  addWord() {
    const newWord = {
      name: this.state.addWordName,
      definition: this.state.addWordDef
    }
    const list = [...this.state.words, newWord];
    this.setState({words: list, addWordName: '', addWordDef: ''});
  }

  render() {
    const words = this.state.words;
    const index = this.state.index;
    const hidden = this.state.hideMeaning;

    return (
      <div>
      <div className="navigation">
        <nav>
          <a href="/">Add Word</a>
          |<a href="">Test</a>
          |<a href="">Word List</a>
        </nav>
      </div>
        <div>Word: {words[index].name}</div>
        <Meaning definition={words[index].definition}
          hidden={hidden} onClick={this.showDefinition}
        />
        <div className="interactions">
          <button onClick={this.getPreviousWord}>Prev</button>
          <button onClick={this.getNextWord}>Next</button>
        </div>
        <div>
          <AddWord onClick={this.addWord}
            nameOnChange={this.nameOnChange}
            defOnChange={this.defOnChange}
            name={this.state.addWordName}
            def={this.state.addWordDef}
          />
        </div>
      </div>
    );
  }
}

function Meaning({definition, hidden, onClick}) {
  return (
    <div>
      <span>Meaning:</span>
      {hidden
       ? <button onClick={onClick}>Show</button>
       : <span>{definition}</span>
      }
    </div>
  );
}

function AddWord({onClick, nameOnChange, defOnChange, name, def}) {
  return (
    <div>
      Name:<input type="text"
        onChange={nameOnChange}
        value={name}
      ></input>
      <br />
      Definition:<textarea
        onChange={defOnChange}
        value={def}
      ></textarea>
      <br />
      <button onClick={onClick}>Create</button>
    </div>
  );
}

//TO DO: Component Composition for simple navigation
//TO DO: Component Composition for confirmation
//TO DO: Component Composition for loading

export default App;
