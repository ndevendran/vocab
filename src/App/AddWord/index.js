import 'React' from 'react';
import {InputGroup, Button} from 'react-bootstrap';

function AddWord({onClick, nameOnChange, defOnChange, name, def, hideStep}) {
  return (
    <div className="addWord">
      {hideStep
        ? <div></div>
        :
        <div>
        <InputGroup className="mb-3 wordInput">
          <InputGroup.Prepend>
            <InputGroup.Text>Name</InputGroup.Text>
          </InputGroup.Prepend>
          <FormControl
            onChange={nameOnChange}
            value={name}
          />
        </InputGroup>
          <br />
          <InputGroup className="definitionInput">
            <InputGroup.Prepend>
              <InputGroup.Text>Definition</InputGroup.Text>
            </InputGroup.Prepend>
            <FormControl as="textarea"
              onChange={defOnChange}
              value={def}
            />
          </InputGroup>
          <br />
          <Button variant="primary" onClick={onClick}>Add Word</Button>
        </div>
      }
    </div>
  );
}
