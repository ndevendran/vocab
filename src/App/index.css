.App {
  text-align: center;
  font-size: 100%;
}

.App-logo {
  height: 40vmin;
}

.interactions {
  margin: auto;
  width: 20%;
  display: flex;
  justify-content: space-between;
  margin-top: 2.5%;
}

.navigation {
  padding-top: 2.5%;
  margin: auto;
  width: 40%;
  display: flex;
  justify-content: space-between;
}

.word {
  margin-bottom: 2.5%;
  margin-top: 2.5%;
}

.wordInput {
  width: 25% !important;
  margin: auto;
  margin-top: 2.5%;
}

.definitionInput {
  width: 40% !important;
  margin: auto;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}
