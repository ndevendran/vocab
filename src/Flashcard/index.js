import React, {Component} from 'react';
import {Button} from 'react-bootstrap';
import Meaning from '../Meaning';
import PropTypes from 'prop-types';

Button.propTypes = {
  onClick: PropTypes.func
};

Meaning.propTypes = {
  onClick: PropTypes.func,
  definition: PropTypes.string,
  hidden: PropTypes.bool
};


class Flashcard extends Component {
  constructor(props) {
    super(props);

    this.state = {
      index: 0,
      hideMeaning: true,
    };

    this.getPreviousWord = this.getPreviousWord.bind(this);
    this.getNextWord = this.getNextWord.bind(this);
    this.showDefinition = this.showDefinition.bind(this);
  }

  showDefinition() {
    const hideMeaning = !this.state.hideMeaning;
    this.setState({hideMeaning: hideMeaning});
  }

  getNextWord() {
      this.setState((prevState) => {
        let newIndex;
        if((prevState.index + 1) === this.props.words.length) {
          newIndex = 0;
        } else {
          newIndex = prevState.index+1;
        }
        return {index: newIndex, hideMeaning: true};
      });
  }

  getPreviousWord() {
    this.setState((prevState) => {
      let newIndex;
      if(prevState.index === 0) {
        newIndex = this.props.words.length - 1;
      } else {
        newIndex = prevState.index - 1;
      }
      return {index: newIndex, hideMeaning: true};
    });
  }

  render () {
    const hideMeaning = this.state.hideMeaning;
    const word = this.props.words[this.state.index];
    return (
      <div className="flashcard">

          <div>
            <div className="word">Word: {word.name}</div>
            <Meaning definition={word.definition}
              hidden={hideMeaning} onClick={this.showDefinition}
            />
            <div className="interactions">
              <Button variant="primary" onClick={this.getPreviousWord}>Prev</Button>
              <Button variant="primary" onClick={this.getNextWord}>Next</Button>
            </div>
        </div>
      </div>
    );
  }
}

export default Flashcard;
