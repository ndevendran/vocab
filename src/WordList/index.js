import React from 'react';
import {Button} from 'react-bootstrap';
import './index.css';

export default function WordList({words, dismiss, onDelete}) {
  const list = words || [];
  return (
    <div className="wordList">
        <table className="wordList">
          <tbody>
          {list.map((word) => {
            return (
              <tr key={word.name}>
                <td>{word.name}</td>
                <td>
                <Button variant="secondary"
                  className="delete"
                  onClick={() => onDelete(word.name)}
                >
                  Delete
                </Button>
                </td>
              </tr>
            );
          })}
          </tbody>
        </table>
    </div>
  );
}
