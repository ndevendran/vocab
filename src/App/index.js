import React, {Component} from 'react';
import './index.css';
import AddWord from '../AddWord';
import Flashcard from '../Flashcard';
import WordList from '../WordList';
import {Nav} from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import PropTypes from 'prop-types';

const list = [
  { name: "malevolent",
    definition: "having or showing a wish to do evil to others"},
  { name: "pejorative",
    definition: "expressing contempt or disapproval"}
];

const FLASHCARD = "flashcard";
const DEFAULT_STEP = "flashcard";
const ADD_WORD = "add";
const WORD_LIST = "list";

AddWord.propTypes = {
  onClick: PropTypes.func,
  hideStep: PropTypes.bool
};

Flashcard.propTypes = {
  words: PropTypes.array,
  index: PropTypes.number,
  hideStep: PropTypes.bool,
  hideMeaning: PropTypes.bool
};

WordList.propTypes = {
  words: PropTypes.array,
  hideStep: PropTypes.bool
};

Nav.Item.propTypes = {
  onClick: PropTypes.func
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      words: list,
      index: 0,
      hideMeaning: true,
      addWordName: '',
      addWordDef: '',
      currentStep: DEFAULT_STEP,
    }

    this.addWord = this.addWord.bind(this);
    this.setStep = this.setStep.bind(this);
  }

  hideStep(step) {
    return !(step === this.state.currentStep);
  }

  setStep(step) {
    this.setState({currentStep: step});
  }

  addWord(name, def) {
    const newWord = {name: name, definition: def};
    const duplicate = this.state.words.filter(word => word.name === newWord.name);

    if(duplicate.length !== 0) {
      alert("This word already exists. If you want to redefine it, delete the word from the word list first");
    } else {
      const list = [...this.state.words, newWord];
      this.setState({words: list});
    }
  }

  render() {
    const words = this.state.words;
    const index = this.state.index;
    const hidden = this.state.hideMeaning;

    return (
      <div className="App">
          <Nav variant="tabs" className="navigation">
            <Nav.Item onClick={() => this.setStep(ADD_WORD)}><Nav.Link>Add Word</Nav.Link></Nav.Item>
            <Nav.Item href="" onClick={() => this.setStep(FLASHCARD)}><Nav.Link>Flashcards</Nav.Link></Nav.Item>
            <Nav.Item href="" onClick={() => this.setStep(WORD_LIST)}><Nav.Link>Word List</Nav.Link></Nav.Item>
          </Nav>
          <div>
            <AddWord onClick={this.addWord}
              hideStep={this.hideStep(ADD_WORD)}
            />
          </div>
          <div>
            <Flashcard
              words={words}
              index={index}
              hideStep={this.hideStep(FLASHCARD)}
              hideMeaning={hidden}
            />
          </div>
          <div>
            <WordList hideStep={this.hideStep(WORD_LIST)}
              words={this.state.words}
             />
          </div>
      </div>
    );
  }
}

//TO DO: Component Composition for simple navigation
//TO DO: Component Composition for confirmation
//TO DO: Component Composition for loading
//TO DO: Class Composition with persistence layer
//TO DO: Custom Button, Text Area, Input, and Navigation
//TO DO: Delete Word

export default App;
