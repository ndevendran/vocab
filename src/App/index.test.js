import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../App';
import Flashcard from '../Flashcard';
import AddWord from '../AddWord';
import WordList from '../WordList';

Enzyme.configure({adapter: new Adapter()});

describe('App', () => {
  const props = {
    list: [{name: "test1", definition: "test"}]
  };

  it('loads without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has navigation to add words', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find('#addWord')).toHaveLength(1);
  });

  it('has navigation to flashcards', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find('#flashcards')).toHaveLength(1);
  });

  it('has navigation to the word list', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find('#wordList')).toHaveLength(1);
  });

  it('should see add flashcards by default', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find(Flashcard)).toHaveLength(1);
  });

  it('should change to add card when add card nav is clicked', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find(AddWord)).toHaveLength(0);
    wrapper.find('#addWord').simulate('click');
    expect(wrapper.find(AddWord)).toHaveLength(1);
    expect(wrapper.find(Flashcard)).toHaveLength(0);
    expect(wrapper.find(WordList)).toHaveLength(0);
  });

  it('should change to word list when word list nav is clicked', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find(WordList)).toHaveLength(0);
    wrapper.find('#wordList').simulate('click');
    expect(wrapper.find(WordList)).toHaveLength(1);
    expect(wrapper.find(AddWord)).toHaveLength(0);
    expect(wrapper.find(Flashcard)).toHaveLength(0);
  });

  it('should change back to flashcard when we navigate away and back', () => {
    const wrapper = shallow(<App {...props} />);
    wrapper.find('#wordList').simulate('click');
    expect(wrapper.find(Flashcard)).toHaveLength(0);
    wrapper.find('#flashcards').simulate('click');
    expect(wrapper.find(Flashcard)).toHaveLength(1);
    expect(wrapper.find(WordList)).toHaveLength(0);
    expect(wrapper.find(AddWord)).toHaveLength(0);
  });

  //it should remove word from list when delete is clicked
  //it should add word to list when input is filled out and add is clicked
  //added word should appear in word list
})
