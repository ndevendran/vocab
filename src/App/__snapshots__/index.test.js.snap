// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="App"
>
  <div
    className="navigation nav nav-tabs"
    onKeyDown={[Function]}
  >
    <div
      className="nav-item"
      onClick={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Add Word
      </a>
    </div>
    <div
      className="nav-item"
      href=""
      onClick={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Flashcards
      </a>
    </div>
    <div
      className="nav-item"
      href=""
      onClick={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Word List
      </a>
    </div>
  </div>
  <div>
    <div
      className="addWord"
    >
      <div />
    </div>
  </div>
  <div>
    <div
      className="flashcard"
    >
      <div>
        <div
          className="word"
        >
          Word: 
          malevolent
        </div>
        <div>
          <button
            className="showDefinition btn btn-info"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Show Definition
          </button>
        </div>
        <div
          className="interactions"
        >
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Prev
          </button>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="wordList"
    >
      <div />
    </div>
  </div>
</div>
`;
